### PR TITLE
Add easy cooldown support to script commands

### PR DESCRIPTION
**Pull request by xXAndrew28Xx**
Target Minecraft versions: Not dependent on Minecraft features.
Requirements: None
Related issues: None

Description:
Adds support for specifying cooldowns for script commands, example:
```
on command /home:
    cooldown: 10 seconds
    cooldown message: You may only teleport home every 10 seconds
    trigger:
        teleport player to {home::%uuid of player%}
```
The cooldown accepts a timespan. The cooldown message is optional and defaults to `You are using this command too often, please try again later` for English and `Sie verwenden diesen Befehl zu oft, versuchen Sie es später erneut` for German (Google Translate sadly). 
The cooldown also can be retrieved from inside the trigger of the command (`[the] cooldown [(of [the] command)]`).

### CODE REVIEW

File: src/main/java/ch/njol/skript/command/ScriptCommand.java [hunk 1]
```diff
@@ -22,16 +22,11 @@
...
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (Snow-Pyon) [inline src/main/java/ch/njol/skript/command/ScriptCommand.java]: star import, Skript doesn't use them.

**Comment by xXAndrew28Xx** [comment]
@Snow-Pyon Changes have been made.

**Comment by Blueyescat** [comment]
I like the idea, but i think there should be something about remaining time.

**Comment by xXAndrew28Xx** [comment]
I could possibly do that, but then I would have to parse the provided string as a VariableString. I'll try right now.

**Comment by xXAndrew28Xx** [comment]
I just finished it. Once this PR gets merged, I'll make another PR allowing the same for permission messages. :)

```vb
command /cooldown:
    cooldown: 10 seconds
    cooldown message: &eYou last used this command %elapsed time% ago. You may use it again in %remaining time% as you may only use this command every %cooldown%, %player%.
    trigger:
        message "&aAyy, OK"
```

**Comment by TonyMaster21** [comment]
This is actually awesome... Would save so much code

**Comment by Snow-Pyon** [comment]
Nice work done here, if @bensku doesn't have any complains, I'm merging it.

**Comment by Blueyescat** [comment]
Awesome :blue_heart:

**Comment by bensku** [comment]
Looks good. Configurable permission to skip cooldown would be nice, though.

**Comment by Blueyescat** [comment]
^ Also if a player gets the cooldown permission when the player has the cooldown already it should delete the cooldown and run the trigger.

By the way, what happens when the server is stopped? Or long delays like 1 hours?

**Comment by xXAndrew28Xx** [comment]
@bensku Would `cooldown bypass` be a good name for the node.
@blueyescat I’ll implement that first suggestion. The last usage date is tracked in memory. So any server restarts will reset it. Long delays will work as long as the server isn’t restarted (could I piggyback off of the variable storage to fix this?). Also the cooldowns depends on timespans which have a maximum limit of 29 days IIRC.

**Comment by Blueyescat** [comment]
Hmm why? It can work by saving used date and checking difference between that and now. But yeah, it would reset when the server is stopped. It must be saved at somewhere... Interesting, simple cooldown system is fine but cooldown can be used on other systems.

**Comment by xXAndrew28Xx** [comment]
All suggestions have been implemented besides the cooldown being persistent across restarts, as I'm not sure how the best way to implement that is.

**Comment by xXAndrew28Xx** [comment]
Egh looks like some .idea changes made it somehow.

**Comment by Pikachu920** [comment]
I think it should be documented this isn't suitable for longer cooldowns

**Comment by Syst3ms** [comment]
Unless he adds a way to make it so you can do it without a wait, like how you would do a good cooldown in skript, with timespan variables. That would be ideal.

**Comment by bensku** [comment]
Merging this since @Snow-Pyon approved the code. I'll need to think about how we document this feature at all; currently docs do not contain information about declaring custom commands at all.

**Comment by TheBentoBox** [comment]
It would probably be best to add a tutorials page to the documentation with links to specific tutorials (either hosted on your docs or even ones on the forums or whatever), which could include a tutorial on custom commands.

**Comment by bensku** [comment]
Hmm, not bad idea. Also a separate page for commands, probably, since they have NO docs there at all - not even syntaxes.

**Comment by Blueyescat** [comment]
@xXAndrew28Xx The elapsed time expression returns the remaining time, the remaining time expression returns the elapsed time

File: src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        if (scriptCommand.getCooldown() == null || !(sender instanceof Player)) {
+            return null;
+        }
+        Player player = (Player) event.getSender();
+        UUID uuid = player.getUniqueId();
+
+        switch (mark) {
+            case 0:
+            case 1:
+                long ms = mark == 1
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: What a strange way to use parse marks

File: src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        ? scriptCommand.getRemainingMilliseconds(uuid)
+                        : scriptCommand.getElapsedMilliseconds(uuid);
+                return new Timespan[]{new Timespan(ms)};
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: What a strange way to use parse marks

Comment (Blueyescat) [reply src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: Actually the bug is right there, doesn't it check for if the mark is 1, after another check for if the mark is 1?

Comment (bensku) [reply src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: Switch statements fall through in Java. This check happens if mark is 0 or 1.

Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: Fixed in the latest commit

Comment (xXAndrew28Xx) [reply src/main/java/ch/njol/skript/expressions/ExprCmdElapsedRemainingTime.java]: Oh the order was backwards, oops.

**Comment by melinstagibson** [comment]
Ild also suggest a way to cancel/reset cooldowns or set the remaining time manually within the command.

like now if you take the comand /nick for example there are a few other things a player must not do like using too many characters and you dont want the cooldown to trigger in those cases.

also how do i define the cooldown bypass? i tried everything i could think of

**Comment by Syst3ms** [comment]
It's a permission

**Pull request merged**